README.md
# Legacy Reactor

[![Commitizen friendly](https://img.shields.io/badge/commitizen-friendly-brightgreen.svg)](http://commitizen.github.io/cz-cli/)
[![Build Status](https://travis-ci.org/michael-stoltz/legacy-reactor.svg?branch=master)](https://travis-ci.org/michael-stoltz/legacy-reactor)
[![Coverage Status](https://coveralls.io/repos/github/michael-stoltz/legacy-reactor/badge.svg?branch=master)](https://coveralls.io/github/michael-stoltz/legacy-reactor?branch=master)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
[![docs: typedoc](https://img.shields.io/badge/docs-typedoc-blue.svg)](https://michael-stoltz.github.io/legacy-reactor/)
[![Greenkeeper badge](https://badges.greenkeeper.io/michael-stoltz/legacy-reactor.svg)](https://greenkeeper.io/)

The Legacy Reactor is a standalone implementation of the [vue](https://vuejs.org/) 2.x observer code for reactivity.

The target environment is ES5+ and will therefore work on any browser that supports it.

If you are working with evergreen browsers that support [proxies](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy), rather use the latest [version](https://github.com/michael-stoltz/reactor) of the library.

# Recommended Editor

It is highly suggested to use [Visual Studio Code](https://code.visualstudio.com/) as it is extendible and light weight.

The project comes configured with some settings that make development within vscode as simple as possible.

These include things such as code format on save and debugging current ts files and tests.

# Npm Scripts

- `npm run lint`: Lint all typescript files.
- `npm run format`: Format all ts files in the `src` folder.
- `npm run commit`: Commit using [commitzen](https://github.com/commitizen/cz-cli) to ensure commits follow the correct convention.
- `npm run test`: Run all [jest](https://jestjs.io/) tests on files that end with `*.spec.ts` in the `src` folder.
- `npm run cov`: Generate and open a test coverage report using [jest](https://jestjs.io/).
- `npm run cov:open`: Open the generated test coverage report.
- `npm run doc`: Generate and open documentation using [typedoc](https://typedoc.org/).
- `npm run doc:open`: Open generated documentation.
- `npm run build`: Build the project using the typescript compiler.
- `npm run version`: Generate or update a CHANGELOG.md, bump the package version and create a tag using [Standard Version](https://github.com/conventional-changelog/standard-version).
- `npm run prepublish`: Automatically runs before the `npm publish` command and ensures that the code is formatted, tested and built before publishing.

# Usage

## Computed properties

Computed properties are properties that get their value from other properties.

When you change one of the properties it relies on it will automatically be recalculated.

```javascript
import { observe } from '@microlibs/legacy-reactor';

const observed = observe({
  price: 55.6,
  qty: 100,
  // Any function in this object is treated as a computed property definition.
  total() {
    return this.price * this.total;
  },
});

console.log(observed.total); // output: 5560

observed.price = 60;
console.log(observed.total); // output: 6000

observed.qty = 50;
console.log(observed.total); // output: 3000
```

### Side Effects

Computed properties disable data updates within them to keep your data predictable.

Setting the value of data within a computed property will cause an exception.

```typescript
const observed = observe({
  price: 10,
  qty: 5,
  total() {
    // okay
    var value = this.price * this.total;

    // this line will throw an exception
    this.price = 1000;

    return value;
  },
});
```

## Watchers

Watchers are simple functions that get run when the value of an observed property changes.

You can register multiple watchers per property.

```javascript
import { observe, addPropertyWatcher, removePropertyWatcher } from '@microlibs/legacy-reactor';

const observed = observe({
  price: 55.6,
  qty: 100,
  // Any function in this object is treated as a computed property definition.
  total() {
    return this.price * this.total;
  },
});

// The first parameter is an observed object
// The second parameter is the path to the property in the observed data.
// In case of a nested object {nested: {total: 55 }} you would write 'nested.total'.
// The last parameter is the function to be called when the value of total changes.
const watcher = addPropertyWatcher(observed, 'total', (value, oldValue) => {
  console.log(value, oldValue);
});

observed.price = 100; // output: 10000 5560

removePropertyWatcher(observed, 'total', watcher);

// No output since the watcher was removed
observed.price = 60.56;
```

# Known Issues

Since the reactor uses the same reactivity mechanism as vue does it comes with the same caveats.

1. Adding properties to an observed object is not tracked.
2. Deleting properties from an observed object is not tracked.
   > Due to the above two caveats we decided to seal the observed object.
   > Therefore you cannot add or remove properties once you have created an observed object.
3. Reassigning array length is not tracked.
   > It is not possible to define a getter on the length property since it is not configurable.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [0.5.0](https://github.com/michael-stoltz/legacy-reactor/compare/v0.4.0...v0.5.0) (2019-07-19)


### Bug Fixes

* disable side effects in user defined getters ([ca263cc](https://github.com/michael-stoltz/legacy-reactor/commit/ca263cc))


### Features

* add support for user defined getters in observed data ([4583030](https://github.com/michael-stoltz/legacy-reactor/commit/4583030))



## [0.4.0](https://github.com/michael-stoltz/legacy-reactor/compare/v0.3.0...v0.4.0) (2019-07-17)


### Features

* computed properties can no longer have side effects ([eee66eb](https://github.com/michael-stoltz/legacy-reactor/commit/eee66eb))



## [0.3.0](https://github.com/michael-stoltz/legacy-reactor/compare/v0.2.2...v0.3.0) (2019-07-12)


### Build System

* change travis ci config ([e253ef6](https://github.com/michael-stoltz/legacy-reactor/commit/e253ef6))


### Features

* add travis ci and coveralls support ([3a92f78](https://github.com/michael-stoltz/legacy-reactor/commit/3a92f78))


### Refactor

* Change the add and remove watcher api ([#7](https://github.com/michael-stoltz/legacy-reactor/issues/7)) ([86cb544](https://github.com/michael-stoltz/legacy-reactor/commit/86cb544)), closes [#2](https://github.com/michael-stoltz/legacy-reactor/issues/2)


### BREAKING CHANGES

* observed data no longer has $watch and $unwatch functions

* refactor: remove remnants of the old watcher mechanism

* fix: return the watcher function from addPropertyWatcher



### 0.2.2 (2019-07-03)



### 0.2.1 (2019-07-03)


src/observer/index.ts
/**
 * Module with functions that emulate [vue's](https://vuejs.org/) reactivity mechanism.
 */

/** @ignore */
import { isObject, isPlainObject, prototypeAugment } from '../util';
import { arrayMethods } from './array';
import ComputedObservable from './computed-observable';
import Observable, { shouldObservablesUpdate } from './observable';
import { ATTACHED_OBSERVABLE_KEY, ComputedFunction, IObservable, IObservableReference, ObservedData, WatcherFunction } from './types';

/**
 * The [[IObservable]] currently being created and evaluated.
 *
 * This property is globally unique because only one [[IObservable]] can be evaluated at a time.
 */
let currentEvaluatingObservable: IObservable<any> | undefined;

/**
 * Takes a data object and recursively makes all its properties reactive.
 *
 * ## Computed Properties
 *
 * Function definitions within the data object are treated as computed property definitions.
 *
 * ```typescript
 * const observed = observe({
 *  price: 55,
 *  quantity: 10,
 *  total() {
 *    return this.price * this.quantity;
 *  }
 * });
 *
 * console.log(observed); // output: { price: 55, quantity: 10, total: 550 }
 * ```
 *
 * @param data - Object to process.
 *
 * @typeparam T - Plain javascript object.
 */
export function observe<T extends object>(data: T): ObservedData<T> {
  if (isPlainObject(data)) {
    observeObject(data as T);
  } else {
    throw new Error('Parameter provided is not a plain javascript object.');
  }

  return data as ObservedData<T>;
}

/**
 * Iterate over a data object and make all its properties reactive.
 *
 * @param data - Data object.
 * @param observable - Observable for the data object
 *
 * @typeparam T - Any object type: array, object, class etc.
 */
export function observeObject<T extends object>(data: T, observable?: IObservable<T>): void {
  if (isObject(data)) {
    let shouldSeal = true;

    if (Array.isArray(data)) {
      shouldSeal = false;
      if (!(data as IObservableReference<T>).__observable__) {
        Object.defineProperty(data, ATTACHED_OBSERVABLE_KEY, { value: observable });
        prototypeAugment((data as unknown) as object, arrayMethods);
      }
    }

    for (const key in data) {
      if (data.hasOwnProperty(key)) {
        let value = data[key];
        let valueObservable: IObservable<typeof value>;

        if (typeof value === 'function') {
          value = value.bind(data);
          valueObservable = new ComputedObservable((value as unknown) as ComputedFunction<typeof value>);

          currentEvaluatingObservable = valueObservable;
          const evaluatedObservableValue = (valueObservable as ComputedObservable<T>).evaluate()!;
          valueObservable.update((evaluatedObservableValue as unknown) as T[Extract<keyof T, string>]);
          currentEvaluatingObservable = undefined;
        } else {
          valueObservable = new Observable(value);

          observeObject((value as unknown) as T, (valueObservable as unknown) as IObservable<T>);
        }

        defineReactiveProperty(data, key, valueObservable);
      }
    }

    if (shouldSeal) {
      Object.seal(data);
    }
  }
}

/**
 * Creates a reactive property on a specified object.
 *
 * For a property to be considered reactive it needs to be proxied with a getter/setter and also have an associated [[Observable]] instance.
 *
 * ### Reactive properties
 * ```typescript
 * const obj = {};
 * defineReactiveProperty(obj, 'number', new Observable(99));
 *
 * // Note that even though the value is proxied you can still access it as you normally access properties.
 * console.log(obj.number) // output: 99
 * obj.number = 105;
 * console.log(obj.number) // output: 105
 * ```
 *
 * @param obj - Object on which to create the reactive property.
 * @param key - Key for the new property.
 * @param observable - [[Observable]] instance that stores the value of the reactive property.
 *
 * @typeparam T - Any valid javascript value.
 */
export function defineReactiveProperty<T>(obj: object, key: string | number, observable: IObservable<T>) {
  const propertyDescriptor = Object.getOwnPropertyDescriptor(obj, key);
  const getter = propertyDescriptor ? propertyDescriptor.get : undefined;
  const setter = propertyDescriptor ? propertyDescriptor.set : undefined;

  Object.defineProperty(obj, key, {
    get() {
      if (arguments[0] === true) {
        return observable;
      } else {
        if (currentEvaluatingObservable) {
          observable.observe(currentEvaluatingObservable as ComputedObservable<T>);
        }
        shouldObservablesUpdate(false);
        const value = getter ? getter.call(obj) : observable.value;
        shouldObservablesUpdate(true);
        return value;
      }
    },
    // prettier-ignore
    set: observable instanceof ComputedObservable ? () => { /* */ } : function (newValue: T) {      
      if (setter) {
        setter.call(obj, newValue);
      }
      newValue = getter ? getter.call(obj): newValue;
      if (newValue !== observable.value) {
        observeObject(newValue as unknown as object, observable as unknown as IObservable<object>);
        observable.update(newValue);
      }
    },
    enumerable: true,
  });
}

/**
 * Extracts the [[Observable]] instance from a property on an object.
 *
 * This function will only work if the [[defineReactiveProperty]] method was used to define that property.
 *
 * @param object - Object where you have a reactive property.
 * @param key - Key of the property that has an observable instance.
 */
export function extractObservableFromProperty(object: object, key: string | number): IObservable<any> | undefined {
  const propertyDescriptor: PropertyDescriptor | undefined = Object.getOwnPropertyDescriptor(object, key);
  if (propertyDescriptor && propertyDescriptor.get) {
    return (propertyDescriptor.get as any)(true);
  }
  return undefined;
}

/**
 * Traverses a string path for an object to find the property it is pointing to.
 *
 * Once the property is found a callback function is called passing in the property's parent object and the property name.
 *
 * ```typescript
 * const data = {
 *   nested: {
 *    anotherNested: {
 *      property: 'value',
 *      number: 66
 *    }
 *   }
 * };
 *
 * navigateToPropertyPath(data, 'nested.anotherNested.property', (obj, property) => {
 *  console.log(obj, property);
 * });
 *
 * // output: {property: 'value', number: 66} property
 * ```
 *
 * @param obj - Object to traverse.
 * @param path - Path to a property on the obj.
 * @param callback - Callback to be called when the property is found.
 */
export function navigateToPropertyPath<T extends object>(obj: T, path: string, callback: (obj: object, key: string) => void): void {
  const properties = path.split('.');
  let property!: string;

  for (let i = 0; i < properties.length; i++) {
    if (obj.hasOwnProperty(properties[i])) {
      property = properties[i];
      if (i !== properties.length - 1) {
        obj = obj[property as keyof object];
      }
    } else {
      throw new Error(`Object does not contain the property with path ${path}`);
    }
  }

  callback(obj, property);
}

/**
 * Finds the observable attached to a property within observed data and adds or removes a watcher from its watcher list.
 *
 * @param observedData - Object containing observed data created by [[observe]].
 * @param path - Path to the property in an object.
 * @param watcher - [[WatcherFunction]].
 * @param operation - Specifies what to do with the [[WatcherFunction]]
 */
function modifyPropertyWatcherList<T extends object>(observedData: T, path: string, watcher: WatcherFunction<any>, operation: 'add' | 'remove') {
  navigateToPropertyPath(observedData, path, (obj, property) => {
    const observable = extractObservableFromProperty(obj, property);

    if (observable) {
      if (operation === 'add') {
        observable.watch(watcher);
      } else if (operation === 'remove') {
        observable.unwatch(watcher);
      }
    } else {
      throw new Error('Property is not an observable property.');
    }
  });
}

/**
 * Adds a watcher function to a property that gets called when the property changes.
 *
 * ```typescript
 * const observed = observe({
 *  price: 43,
 *  qty: 10,
 *  total() {
 *    return this.qty * this.price;
 *  }
 * });
 *
 * addPropertyWatcher(observed, 'price', (value, oldValue) => {
 *  console.log(value, oldValue);
 * });
 *
 * // watcher is called on data change
 * observed.price = 50; // output: 50 43
 * ```
 *
 * @param data - Object observed with [[observe]].
 * @param path - Path to the property on the data object.
 * @param watcher - Function to add to the properties' watchers.
 */
export function addPropertyWatcher<T>(data: object, path: string, watcher: WatcherFunction<T>) {
  modifyPropertyWatcherList(data, path, watcher, 'add');

  return watcher;
}

/**
 * Removes a watcher function from a property.
 *
 * ```typescript
 * const observed = observe({
 *  price: 43,
 *  qty: 10,
 *  total() {
 *    return this.qty * this.price;
 *  }
 * });
 *
 * const watcher = (value, oldValue) => {
 *  console.log(value, oldValue);
 * }
 *
 * addPropertyWatcher(observed, 'price', watcher);
 *
 * // watcher is called on data change
 * observed.price = 50; // output: 50 43
 *
 * removePropertyWatcher(observed, 'price', watcher);
 *
 * // no output since watcher was removed
 * observed.price = 90;
 * ```
 * @param data - Object observed with [[observe]].
 * @param path - Path to the property on the data object.
 * @param watcher - Function to remove from the properties' watchers.
 */
export function removePropertyWatcher<T>(data: object, path: string, watcher: WatcherFunction<T>) {
  modifyPropertyWatcherList(data, path, watcher, 'remove');
}
